velox/functions/prestosql/window/FirstLastValue.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::window::prestosql {
+
+namespace {
+
+enum class ValueType {
+  kFirst,
+  kLast,
+};
+
+template <ValueType TValue, typename T>
+class FirstLastValueFunction : public exec::WindowFunction {
+ public:
+  explicit FirstLastValueFunction(
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool, nullptr) {
+    VELOX_CHECK_NULL(args[0].constantValue);
+    argIndex_ = args[0].index.value();
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) override {
+    partition_ = partition;
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      const SelectivityVector& validRows,
+      int32_t resultOffset,
+      const VectorPtr& result) override {
+    auto numRows = frameStarts->size() / sizeof(vector_size_t);
+    rowNumbers_.resize(numRows);
+
+    if constexpr (TValue == ValueType::kFirst) {
+      auto frameStartsPtr = frameStarts->as<vector_size_t>();
```

Comment:
naming: frameStartsPtr -> rawFrameStarts and rawFrameEnds

Same for NthValueFunction.